### PR TITLE
feat: enable wasm compilation

### PR DESCRIPTION
**Pull request by vaind**
This updates and replaces #2064 - I suggest you have a look before looking into this PR.

With the changes added in this PR, we can support webassembly compilation without adding an explicit dependency on package:web and thus breaking compatibility with older dart/flutter versions. 

As such, this should not be breaking for anyone. I've added tests for every minor dart version since ~~2.17~~ 2.18 and everything seems to be in order.

There's one caveat: we're losing some points in the new `pana` check that downgrades dependencies fully berofe running dart analyze. The weird thing is, it's supposed to be reproducible with doing `dart pub downgrade` followed by `dart analyze`, but that doesn't work (i.e. it won't fail the same way `pana` does - actually it doesn't show any issues that way). 

In any case, with compilation working fine for all dart versions, I think it's safe to ignore the 20 point reduction for now (until we drop old Flutter version support and can add `package:web` dependency directly).

closes #2064 
closes #2110

### CODE REVIEW

**Comment by ueman** [comment]
> fix: changes min flutter version to 3.13.0 and dart sdk 3.1.0 to be compatible with package:web 0.5.1

That should allow you to get rid of various backwards compatibility code in the Flutter SDK

**Comment by vaind** [comment]
> > fix: changes min flutter version to 3.13.0 and dart sdk 3.1.0 to be compatible with package:web 0.5.1
> 
> That should allow you to get rid of various backwards compatibility code in the Flutter SDK

we're actually not going to do that now. the message was autogenerated from the commits but I've rolled back the change. More history in #2064

**Comment by vaind** [comment]
For future reference, Dart 2.17 tests fail after a couple of tests with 

```
===== CRASH =====
si_signo=Segmentation fault(11), si_code=1, si_addr=0x18
version=2.17.7 (stable) (Mon Aug 22 10:43:32 2022 +0200) on "linux_x64"
pid=1792, thread=1825, isolate_group=main(0x5571f6910800), isolate=test/sentry_measurement_unit_test.dart(0x5571f84e8000)
isolate_instructions=5571f450a460, vm_instructions=5571f450a460
  pc 0x00005571f4bb7c4e fp 0x00007f09780fd150 dart::kernel::ScopeBuilder::LookupVariable(long)+0x1fe
  pc 0x00005571f4bb7a0f fp 0x00007f09780fd170 dart::kernel::ScopeBuilder::VisitVariableGet(long)+0xf
  pc 0x00005571f4bb5960 fp 0x00007f09780fd1b0 dart::kernel::ScopeBuilder::VisitExpression()+0x610
  pc 0x00005571f4bb62b8 fp 0x00007f09780fd280 dart::kernel::ScopeBuilder::VisitConstructor()+0x228
  pc 0x00005571f4bb439d fp 0x00007f09780fd400 dart::kernel::ScopeBuilder::BuildScopes()+0x97d
  pc 0x00005571f476c0e7 fp 0x00007f09780fd740 dart::ParsedFunction::EnsureKernelScopes()+0x37
  pc 0x00005571f4b7dcc3 fp 0x00007f09780fd770 dart::kernel::StreamingFlowGraphBuilder::ParseKernelASTFunction()+0x53
  pc 0x00005571f4b7db21 fp 0x00007f09780fd820 dart::kernel::StreamingFlowGraphBuilder::BuildGraph()+0xb1
  pc 0x00005571f4b916d5 fp 0x00007f09780fdae0 dart::kernel::FlowGraphBuilder::BuildGraph()+0x65
  pc 0x00005571f48070f0 fp 0x00007f09780fdd20 dart::DartCompilationPipeline::BuildFlowGraph(dart::Zone*, dart::ParsedFunction*, dart::ZoneGrowableArray<dart::ICData const*>*, long, bool)+0x40
  pc 0x00005571f4808112 fp 0x00007f09780fe430 dart::CompileParsedFunctionHelper::Compile(dart::CompilationPipeline*)+0x4d2
  pc 0x00005571f4808d8c fp 0x00007f09780fe640 dart+0x21e3d8c
  pc 0x00005571f4809175 fp 0x00007f09780fe680 dart::Compiler::EnsureUnoptimizedCode(dart::Thread*, dart::Function const&)+0xb5
  pc 0x00005571f47e2a64 fp 0x00007f09780fe6d0 dart::SourceReport::VisitFunction(dart::JSONArray*, dart::Function const&)+0x134
  pc 0x00005571f47e2efe fp 0x00007f09780fe760 dart::SourceReport::VisitLibrary(dart::JSONArray*, dart::Library const&)+0x12e
  pc 0x00005571f47e338f fp 0x00007f09780fe800 dart::SourceReport::PrintJSON(dart::JSONStream*, dart::Script const&, dart::TokenPosition, dart::TokenPosition)+0x12f
  pc 0x00005571f47d355f fp 0x00007f09780fe950 dart+0x21ae55f
  pc 0x00005571f47ca2a1 fp 0x00007f09780feaa0 dart::Service::InvokeMethod(dart::Isolate*, dart::Array const&, bool)+0x321
  pc 0x00005571f47ca9f1 fp 0x00007f09780feac0 dart::Service::HandleIsolateMessage(dart::Isolate*, dart::Array const&)+0x11
  pc 0x00005571f46ab5e8 fp 0x00007f09780feba0 dart::IsolateMessageHandler::HandleMessage(std::__2::unique_ptr<dart::Message, std::__2::default_delete<dart::Message> >)+0x278
  pc 0x00005571f46d454d fp 0x00007f09780fec10 dart::MessageHandler::HandleMessages(dart::MonitorLocker*, bool, bool)+0x14d
  pc 0x00005571f46d4c2f fp 0x00007f09780fec60 dart::MessageHandler::TaskCallback()+0x1df
  pc 0x00005571f47f60c8 fp 0x00007f09780fece0 dart::ThreadPool::WorkerLoop(dart::ThreadPool::Worker*)+0x148
  pc 0x00005571f47f651d fp 0x00007f09780fed10 dart::ThreadPool::Worker::Main(unsigned long)+0x6d
  pc 0x00005571f476b368 fp 0x00007f09780fedd0 dart+0x[214](https://github.com/getsentry/sentry-dart/actions/runs/9600075094/job/26475396946?pr=2113#step:3:253)6368
-- End of DumpStackTrace
=== Crash occured when compiling package:sentry/src/sentry_measurement_unit.dart_InformationSentryMeasurementUnit_InformationSentryMeasurementUnit. in unoptimized JIT mode in unknown pass
=== Flow Graph not available
/home/runner/work/_temp/4e36ea00-b559-4259-86cf-09b27b4f401e.sh: line 1:  1792 Aborted                 (core dumped) dart test -p vm --coverage=coverage --test-randomize-ordering-seed=random --chain-stack-traces
```

**Comment by denrase** [comment]
Just tested this locally with `flutter build web --wasm` and our example app (commenting out code that does not work with wasm) . Running the app using `dart pub global run dhttpd '--headers=Cross-Origin-Embedder-Policy=credentialless;Cross-Origin-Opener-Policy=same-origin'` and checking the issues on sentry.io, everything works as expected. 🥳 

Didn't even have to add the `web` package in this test, as we are already adding it implicitly through the `http` dependency. 

Bildschirmfoto 2024-06-24 um 10 42 49

As far as I can tell, this is the best approach, as you have [`described`](https://github.com/getsentry/sentry-dart/pull/2064#issuecomment-2180024061) and for the reasons outlined [`here`](https://github.com/getsentry/sentry-dart/pull/2064#issuecomment-2179945173).

\cc @buenaflor

File: .github/workflows/dart.yml [hunk 1]
```diff
@@ -49,24 +58,14 @@ jobs:
...
-        if: runner.os != 'Windows'
-        run: dart pub global activate webdev
-
       - name: Build example
         working-directory: dart/example
         run: |
           dart pub get
           dart compile aot-snapshot bin/example.dart
 
-      - name: Build Web example
```
Comment (denrase) [inline .github/workflows/dart.yml:62]: Would it make run compile here using `dart compile wasm` and also keep `webdev build` for the `example_web_legacy` package?

File: dart/lib/src/event_processor/enricher/html_enricher_event_processor.dart [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import 'enricher_event_processor.dart';
+
+EnricherEventProcessor enricherEventProcessor(SentryOptions options) {
+  return WebEnricherEventProcessor(
+    html.window,
+    options,
+  );
+}
+
+class WebEnricherEventProcessor implements EnricherEventProcessor {
```
Comment (denrase) [inline dart/lib/src/event_processor/enricher/html_enricher_event_processor.dart:13]: Should be called `HtmlEnricherEventProcessor`?

In any case, the difference between `html` and `web` implementations is only the access to `http.Window/web.Window` properties, the rest is all duplicated code.

Would it make sense to move these properties into a common interface/class and use it based in conditional import?

**Review by denrase — COMMENTED** [review]
Nice, just some questions

**Comment by kevmoo** [comment]
Let us know if we can help here! This is blocking a critical customer! 😄

**Review by buenaflor — APPROVED** [review]
nice 🚀 

didn't see anything major. 

Once all questions are resolved we can go ahead with this 👍 

I think the pana thing is fine for now.

File: .github/workflows/dart.yml (hunk 1 above)
Comment (vaind) [reply .github/workflows/dart.yml:62]: thats done here: https://github.com/getsentry/sentry-dart/pull/2113/files/cec7756d37472e399f4909a557955e76c20c883f#diff-f44f3e514d2ec699f9bf82bd2f0fc5721c20bcdef54a5ebcd7c641b97e39981c

so i've removed the duplicate build directly from the ci

File: dart/lib/src/event_processor/enricher/html_enricher_event_processor.dart (hunk 2 above)
Comment (vaind) [reply dart/lib/src/event_processor/enricher/html_enricher_event_processor.dart:13]: > Should be called `HtmlEnricherEventProcessor`?

The name is asserted in tests so I kept it the same
```dart

    test('$WebEnricherEventProcessor gets added on init', () async {
      late SentryOptions sentryOptions;
      await Sentry.init(
        (options) {
          options.dsn = fakeDsn;
          sentryOptions = options;
        },
      );
      await Sentry.close();

      expect(sentryOptions.eventProcessors.map((e) => e.runtimeType.toString()),
          contains('$WebEnricherEventProcessor'));
    });
```


> In any case, the difference between `html` and `web` implementations is only the access to `http.Window/web.Window` properties, the rest is all duplicated code.
> 
> Would it make sense to move these properties into a common interface/class and use it based in conditional import?

I started doing that last week but then noticed there are more changes in actual implementations so I don't think it makes much sense to extract the common functions. See diff:
```diff
1c1,3
< import 'dart:html' as html show window, Window;
---
> // We would lose compatibility with old dart versions by adding web to pubspec.
> // ignore: depend_on_referenced_packages
> import 'package:web/web.dart' as web show window, Window, Navigator;
8c10
<     html.window,
---
>     web.window,
19c21
<   final html.Window _window;
---
>   final web.Window _window;
62,65c64,66
<       screenHeightPixels: device?.screenHeightPixels ??
<           _window.screen?.available.height.toInt(),
<       screenWidthPixels:
<           device?.screenWidthPixels ?? _window.screen?.available.width.toInt(),
---
>       screenHeightPixels:
>           device?.screenHeightPixels ?? _window.screen.availHeight,
>       screenWidthPixels: device?.screenWidthPixels ?? _window.screen.availWidth,
80,87c81,83
<     final screenOrientation = _window.screen?.orientation;
<     if (screenOrientation != null) {
<       if (screenOrientation.type?.startsWith('portrait') ?? false) {
<         return SentryOrientation.portrait;
<       }
<       if (screenOrientation.type?.startsWith('landscape') ?? false) {
<         return SentryOrientation.landscape;
<       }
---
>     final screenOrientation = _window.screen.orientation;
>     if (screenOrientation.type.startsWith('portrait')) {
>       return SentryOrientation.portrait;
88a85,87
>     if (screenOrientation.type.startsWith('landscape')) {
>       return SentryOrientation.landscape;
>     }
103a103,106
> 
> extension on web.Navigator {
>   external double? get deviceMemory;
> }
```

Comment (denrase) [reply dart/lib/src/event_processor/enricher/html_enricher_event_processor.dart:13]: Makes sense, no reason then to do this. 👍

**Review by denrase — APPROVED** [review]
Nice, thx for doing this! \o/

**Pull request merged**